### ISSUE TITLE
Time Picker value different in android and iOS 

### ISSUE THREAD

**Issue by mehtahardikr** [flutter/flutter#83714]
## Steps to Reproduce

```
final TimeOfDay result =  await showTimePicker(
     context: context,
     initialTime: TimeOfDay(hour: 7, minute: 15),
     builder: (BuildContext context, _) {
       return MediaQuery(
         data: MediaQuery.of(context).copyWith(alwaysUse24HourFormat: false),
         child: _,
       );
     },
   );
   if (result != null) {
     print('Selected time: ${result.format(context)}');
   }
```

**Expected results:** 
it should return same result of  `result.format(context)`
it returns time in 24 hours format.

**Actual results:** 
for iOS returns in 12 hours format
for android returns in 24 hours format [device timeformat doesn't matter]

<details>
  <summary>Logs</summary>

```
[✓] Flutter (Channel stable, 2.2.1, on macOS 11.0.1 20B29 darwin-x64, locale en-GB)
    • Flutter version 2.2.1 at /Users/meghdoot/Library/Android/flutter
    • Framework revision 02c026b03c (4 days ago), 2021-05-27 12:24:44 -0700
    • Engine revision 0fdb562ac8
    • Dart version 2.13.1

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/meghdoot/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.3, Build version 12C33
    • CocoaPods version 1.10.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] VS Code (version 1.55.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.21.0

[✓] Connected device (2 available)
    • iPhone 8 Plus (mobile) • DD25071D-FF64-4C85-89CE-8879154A1E35 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-3 (simulator)
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 90.0.4430.212
```

</details>

**Comment by TahaTesser** [comment]
Hi @mehtahardikr 

> for iOS returns in 12 hours format 
It does always return 12 hours on iOS (same as you)


> for android returns in 24 hours format [device timeformat doesn't matter]
Also return 12 hours, even if the device is on 24 hours

<img src="https://user-images.githubusercontent.com/48603081/120342810-cfe15700-c300-11eb-9c23-e44bb9325037.png" height="450" />	















<details> 
<summary>minimal code sample</summary> 

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Material App',
      theme: ThemeData.dark(),
      home: Home(),
    );
  }
}

class Home extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Material App Bar'),
      ),
      body: Center(
        child: Container(
          child: Text('Hello World'),
        ),
      ),
      floatingActionButton: FloatingActionButton(
        child: Icon(Icons.add),
        onPressed: () async {
          final TimeOfDay result = await showTimePicker(
            context: context,
            initialTime: TimeOfDay(hour: 7, minute: 15),
            builder: (BuildContext context, _) {
              return MediaQuery(
                data: MediaQuery.of(context)
                    .copyWith(alwaysUse24HourFormat: false),
                child: _,
              );
            },
          );
          if (result != null) {
            print('Selected time: ${result.format(context)}');
          }
        },
      ),
    );
  }
}
``` 

</details>

<details> 
<summary>flutter doctor -v</summary> 
	
```console
[✓] Flutter (Channel stable, 2.2.1, on Microsoft Windows [Version 10.0.19043.1023], locale en-US)
    • Flutter version 2.2.1 at C:\Users\Taha\Code\flutter_stable
    • Framework revision 02c026b03c (5 days ago), 2021-05-27 12:24:44 -0700
    • Engine revision 0fdb562ac8
    • Dart version 2.13.1

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at C:\Users\Taha\Code\SDK
    • Platform android-30, build-tools 30.0.3
    • ANDROID_SDK_ROOT = C:\Users\Taha\Code\SDK
    • Java binary at: C:\Users\Taha\Code\openjdk\bin\java.exe
    • Java version OpenJDK Runtime Environment (build 1.8.0-262-b10)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[✓] Visual Studio - develop for Windows (Visual Studio Community 2019 16.10.0)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.10.31321.278
    • Windows 10 SDK version 10.0.19041.0

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/windows#android-setup for detailed instructions).

[✓] IntelliJ IDEA Community Edition (version 2021.1)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2021.1.1
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.56.2)
    • VS Code at C:\Users\Taha\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.22.0

[✓] Connected device (4 available)
    • AOSP on IA Emulator (mobile) • emulator-5554 • android-x86    • Android 9 (API 28) (emulator)
    • Windows (desktop)            • windows       • windows-x64    • Microsoft Windows [Version 10.0.19043.1023]
    • Chrome (web)                 • chrome        • web-javascript • Google Chrome 90.0.4430.212
    • Edge (web)                   • edge          • web-javascript • Microsoft Edge 91.0.864.37

! Doctor found issues in 1 category.
```

</details>

**Comment by mehtahardikr** [comment]
@TahaTesser : but why android has in 24 hours format as i checked in samsung device also.

**Comment by mehtahardikr** [comment]
@TahaTesser  : If you change time to 24 hour format in samsung device. Time picker formats in 24 hour format instead of 12 hour format. t

**Comment by TahaTesser** [comment]
Hi @mehtahardikr 
My bad, I tested on with app kill and updated the https://github.com/flutter/flutter/issues/83714#issuecomment-852183972

As you mentioned, a Samsung device is returning 12-hour format even if the mobile  is set 24 hours as a result i tried a Samsung  device (MO2s) running Android 11, it seems to work correctly below is the result

### iPhone device

| 12-hour | 24-hour |
| --------------- | --------------- |
<img src="https://user-images.githubusercontent.com/48603081/120460697-058a4c80-c3a2-11eb-9a0b-6b9b0916d52b.jpg" height="450" /> | <img src="https://user-images.githubusercontent.com/48603081/120460711-08853d00-c3a2-11eb-8ee6-3e17482661d6.jpg" height="450" /> |


### Samsung device


| 12-hour | 24-hour |
| --------------- | --------------- |
<img src="https://user-images.githubusercontent.com/48603081/120460826-22268480-c3a2-11eb-9f0c-c4d042c2c05a.png" height="450" /> | <img src="https://user-images.githubusercontent.com/48603081/120460833-2357b180-c3a2-11eb-99e8-78079747f026.png" height="450" /> |


<details> 
<summary>minimal code sample</summary> 

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Material App',
      theme: ThemeData.dark(),
      home: Home(),
    );
  }
}

class Home extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Material App Bar'),
      ),
      body: Center(
        child: Container(
          child: Text('Hello World'),
        ),
      ),
      floatingActionButton: FloatingActionButton(
        child: Icon(Icons.add),
        onPressed: () async {
          final TimeOfDay? result = await showTimePicker(
            context: context,
            initialTime: TimeOfDay(hour: 7, minute: 15),
     
          );
          if (result != null) {
            print('Selected time: ${result.format(context)}');
          }
        },
      ),
    );
  }
}

``` 

</details>


<details> 
<summary>flutter doctor -v</summary> 
	
```console
[✓] Flutter (Channel stable, 2.2.1, on Microsoft Windows [Version 10.0.19043.1023], locale en-US)
    • Flutter version 2.2.1 at C:\Users\Taha\Code\flutter_stable
    • Framework revision 02c026b03c (6 days ago), 2021-05-27 12:24:44 -0700
    • Engine revision 0fdb562ac8
    • Dart version 2.13.1

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at C:\Users\Taha\Code\SDK
    • Platform android-30, build-tools 30.0.3
    • ANDROID_SDK_ROOT = C:\Users\Taha\Code\SDK
    • Java binary at: C:\Users\Taha\Code\openjdk\bin\java.exe
    • Java version OpenJDK Runtime Environment (build 1.8.0-262-b10)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[✓] Visual Studio - develop for Windows (Visual Studio Community 2019 16.10.0)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.10.31321.278
    • Windows 10 SDK version 10.0.19041.0

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/windows#android-setup for detailed instructions).

[✓] IntelliJ IDEA Community Edition (version 2021.1)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2021.1.1
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.56.2)
    • VS Code at C:\Users\Taha\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.23.0

[✓] Connected device (5 available)
    • SM M025F (mobile)            • R9ZR205XX0A   • android-arm    • Android 11 (API 30)
    • AOSP on IA Emulator (mobile) • emulator-5554 • android-x86    • Android 9 (API 28) (emulator)
    • Windows (desktop)            • windows       • windows-x64    • Microsoft Windows [Version 10.0.19043.1023]
    • Chrome (web)                 • chrome        • web-javascript • Google Chrome 90.0.4430.212
    • Edge (web)                   • edge          • web-javascript • Microsoft Edge 91.0.864.37

! Doctor found issues in 1 category.
```

</details>


Since you're experiencing this issue on Android 8.1, i tried emulator or 8.1, still no issue
```
I/flutter ( 4927): Selected time: 23:15
```

**Comment by mehtahardikr** [comment]
@TahaTesser  : i am still facing thsi issue with this fucntion : result.format(context) 
as in  Samsung device i am getting 5:00 if device has 24 hours in format
and if device is in 12 hours format then i am getting 5:00  pm. 
the thing is that if you change device time to 24 hour or 12 hours you need to restart app.

and i am using this code 

```
final TimeOfDay result =  await showTimePicker(
     context: context,
     initialTime: TimeOfDay(hour: 7, minute: 15),
     builder: (BuildContext context, Widget child) {
       return MediaQuery(
         data: MediaQuery.of(context).copyWith(alwaysUse24HourFormat: false),
         child: child,
       );
     },
   );
   if (result != null) {
     print('Selected time: ${result.format(context)}');
   }
```

**Comment by mehtahardikr** [comment]
@TahaTesser  : see this screenshot : 
![flutter_01](https://user-images.githubusercontent.com/12496655/120469218-b2bf8d80-c3bf-11eb-8556-507a96d81f82.png)
![flutter_02](https://user-images.githubusercontent.com/12496655/120469234-b81cd800-c3bf-11eb-93c6-d8ddaf81c65c.png)

1. 12 hours timeformat in device 
2. 24 hours timeformat in device

and resuts of `result.format(context)`
1. selected time : 9:40 PM
2. selected time : 21:40

isn't it wrong ?

**Comment by TahaTesser** [comment]
@mehtahardikr 

> I have checked in samsung galaxy device with android 8.1 oreo android o.s.


This could be specific to the Samsung device with Android 8.1 you have. Can you please provide details of this device? Do you have any other Android device that reproduces this issue?

**Comment by mehtahardikr** [comment]
Yes, I have checked on iOS simulator also. by changing region and make device time to 24 hours. 
Its happening in other device of android as well as changing of system time to 24 or 12 requires to app to be kill and reopen again.
`Samsung Galaxy On Max `
`iPhone 6s `
`Samsung Galaxy A72 `

**Comment by TahaTesser** [comment]
> as in Samsung device i am getting 5:00 if device has 24 hours in format
> and if device is in 12 hours format then i am getting 5:00 pm.
> the thing is that if you change device time to 24 hour or 12 hours you need to restart app.

App kill is expected, it is a known issue

So if you kill the app, start it again, you get correct behaviour?

**Comment by mehtahardikr** [comment]
No.

**Comment by TahaTesser** [comment]
@mehtahardikr 
To be clear, which devices have this issue even if you the kill app, you get wrong time format

**Comment by mehtahardikr** [comment]
> @TahaTesser : see this screenshot :
> <img alt="" height="450" src="https://user-images.githubusercontent.com/12496655/120469218-b2bf8d80-c3bf-11eb-8556-507a96d81f82.png"> <img alt="" height="450" src="https://user-images.githubusercontent.com/12496655/120469234-b81cd800-c3bf-11eb-93c6-d8ddaf81c65c.png">
> 
> 1. 12 hours timeformat in device [system time]
> 2. 24 hours timeformat in device [system time]
> 
> and results of `result.format(context)`
> 
> 1. selected time : 9:40 PM
> 2. selected time : 21:40
> 
> isn't it wrong ?

i am saying like that if it is expected behaviour then let me know.

**Comment by TahaTesser** [comment]
Hi @mehtahardikr 
It is not a correct behavior but the devices you mentioned includes iOS device https://github.com/flutter/flutter/issues/83714#issuecomment-853649338, which doesn't have this format issue. I tried the Samsung device myself, it is working fine. You need to kill the app on both platforms, which is expected.

> @mehtahardikr
To be clear, which devices have this issue even if you the kill app, you get the wrong time format

So I am asking which devices exactly which devices have this 24-hour format issue?

**Comment by mehtahardikr** [comment]
@TahaTesser  : 
its on  `Samsung Galaxy OnMax with 8.1 android oreo o.s.`
in `iphone 6s : os. version 12.5.1 , 12.5.2`
in simulator : 8 plus : you can change region to chnage 12 hours to 24 hours from settings -> general ->language & regions
`simluator 8 plus with os. version : 14.4 , 14.2 and 14.3 as well.`

hope this will  help you to check this.

**Comment by TahaTesser** [comment]
Hi @mehtahardikr 

> in iphone 6s : os. version 12.5.1 , 12.5.2

As mentioned, I also tried on iOS, if you kill the app and launch, it works fine


<img width="1440" alt="Screenshot 2021-06-07 at 10 49 34 AM" src="https://user-images.githubusercontent.com/48603081/120979230-10682700-c77e-11eb-871a-97ce6c98b97d.png">

**Comment by mehtahardikr** [comment]
@TahaTesser  : i am not asking about UI , i am askign about out of ` result.format(context)`. see this comment : https://github.com/flutter/flutter/issues/83714#issuecomment-853660276

**Comment by TahaTesser** [comment]
@mehtahardikr 
Working as expected 
### 12-hour
```console
[ +125 ms] Syncing files to device Taha's iPhone...
[   +1 ms] <- reset
[        ] Compiling dart to kernel with 0 updated files
[   +2 ms] <- recompile package:stable_flutter/main.dart a76d373f-3361-43a8-99c6-4f3cd7e6e674
[        ] <- a76d373f-3361-43a8-99c6-4f3cd7e6e674
[ +118 ms] Updating files.
[        ] DevFS: Sync finished
[   +1 ms] Syncing files to device Taha's iPhone... (completed in 124ms)
[        ] Synced 0.0MB.
[   +1 ms] <- accept
[  +15 ms] Connected to _flutterView/0x160034220.
[   +2 ms] Flutter run key commands.
[   +1 ms] r Hot reload. 🔥🔥🔥
[        ] R Hot restart.
[        ] h Repeat this help message.
[        ] d Detach (terminate "flutter run" but leave application running).
[        ] c Clear the screen
[        ] q Quit (terminate the application on the device).
[        ] 💪 Running with sound null safety 💪
[        ] An Observatory debugger and profiler on Taha's iPhone is available at: http://127.0.0.1:55468/Th4_Wf1G_kU=/
[ +547 ms] The Flutter DevTools debugger and profiler on Taha's iPhone is available at:
http://127.0.0.1:9103?uri=http%3A%2F%2F127.0.0.1%3A55468%2FTh4_Wf1G_kU%3D%2F
[+19725 ms] flutter: Selected time: 11:15 AM
```
### 24-hour
```console
[ +122 ms] Syncing files to device Taha's iPhone...
[   +1 ms] <- reset
[        ] Compiling dart to kernel with 0 updated files
[   +5 ms] <- recompile package:stable_flutter/main.dart dd1607f2-eb48-4b50-a9e3-01c09e899ee7
[   +1 ms] <- dd1607f2-eb48-4b50-a9e3-01c09e899ee7
[ +128 ms] Updating files.
[        ] DevFS: Sync finished
[   +1 ms] Syncing files to device Taha's iPhone... (completed in 137ms)
[        ] Synced 0.0MB.
[   +1 ms] <- accept
[  +11 ms] Connected to _flutterView/0x10302ca20.
[   +3 ms] Flutter run key commands.
[   +2 ms] r Hot reload. 🔥🔥🔥
[   +1 ms] R Hot restart.
[        ] h Repeat this help message.
[        ] d Detach (terminate "flutter run" but leave application running).
[        ] c Clear the screen
[        ] q Quit (terminate the application on the device).
[        ] 💪 Running with sound null safety 💪
[        ] An Observatory debugger and profiler on Taha's iPhone is available at: http://127.0.0.1:55545/zeSLMWSwBXI=/
[ +675 ms] The Flutter DevTools debugger and profiler on Taha's iPhone is available at:
http://127.0.0.1:9103?uri=http%3A%2F%2F127.0.0.1%3A55545%2FzeSLMWSwBXI%3D%2F
[+36329 ms] flutter: Selected time: 22:15
```

**Comment by mehtahardikr** [comment]
@TahaTesser  : if timepicker is in 12 hours format then ? it should return 12 hours right rather than  Sytem time is in 24 hours format.

**Comment by TahaTesser** [comment]
Hi @mehtahardikr 
Yes, 12-hour returns this
```console
[   +3 ms] Flutter run key commands.
[   +1 ms] r Hot reload. 🔥🔥🔥
[   +1 ms] R Hot restart.
[        ] h Repeat this help message.
[        ] d Detach (terminate "flutter run" but leave application running).
[        ] c Clear the screen
[        ] q Quit (terminate the application on the device).
[        ] 💪 Running with sound null safety 💪
[        ] An Observatory debugger and profiler on Taha’s iPhone is available at: http://127.0.0.1:53970/vXAw0m9e0u4=/
[ +795 ms] The Flutter DevTools debugger and profiler on Taha’s iPhone is available at:
                    http://127.0.0.1:9103?uri=http%3A%2F%2F127.0.0.1%3A53970%2FvXAw0m9e0u4%3D%2F
[+8086 ms] flutter: Selected time: 11:50 PM
```
24-hour 
```console
[ +122 ms] Syncing files to device Taha's iPhone...
[   +1 ms] <- reset
[        ] Compiling dart to kernel with 0 updated files
[   +5 ms] <- recompile package:stable_flutter/main.dart dd1607f2-eb48-4b50-a9e3-01c09e899ee7
[   +1 ms] <- dd1607f2-eb48-4b50-a9e3-01c09e899ee7
[ +128 ms] Updating files.
[        ] DevFS: Sync finished
[   +1 ms] Syncing files to device Taha's iPhone... (completed in 137ms)
[        ] Synced 0.0MB.
[   +1 ms] <- accept
[  +11 ms] Connected to _flutterView/0x10302ca20.
[   +3 ms] Flutter run key commands.
[   +2 ms] r Hot reload. 🔥🔥🔥
[   +1 ms] R Hot restart.
[        ] h Repeat this help message.
[        ] d Detach (terminate "flutter run" but leave application running).
[        ] c Clear the screen
[        ] q Quit (terminate the application on the device).
[        ] 💪 Running with sound null safety 💪
[        ] An Observatory debugger and profiler on Taha's iPhone is available at: http://127.0.0.1:55545/zeSLMWSwBXI=/
[ +675 ms] The Flutter DevTools debugger and profiler on Taha's iPhone is available at:
http://127.0.0.1:9103?uri=http%3A%2F%2F127.0.0.1%3A55545%2FzeSLMWSwBXI%3D%2F
[+36329 ms] flutter: Selected time: 22:15
```

So far none of the devices/emulator iOS/Android reproduce/ the issue based on your description
Closing, as this isn't an issue with Flutter itself. if you can reproduce the issue consistently with clear steps, please comment below
